contracts/zaps/MintAndBond.sol

```diff
@@ -5,69 +5,159 @@ import "../interfaces/IBondDepository.sol";
-    // 10 PUNK but max bond is 0.32 so you recieve 9.68 PUNK back without paying NFTX fee
-    // ids - 1 < amountToBond < ids
-    require(ids.length.mul(1e18) > amountToBond);
-    require(ids.length.sub(1).mul(1e18) < amountToBond);
+    uint256 _vaultId, 
+    uint256[] calldata _ids, 
+    uint256 _bondId,
+    address _to,
+    uint256 _maxPrice
+  ) external override nonReentrant returns (uint256 remaining_) {
+    require(_to != address(0) && _to != address(this));
+    require(_ids.length > 0);
+
+    (,,,,uint64 maxPayout,,) = bondDepository.markets(_bondId);
+    uint256 bondPrice = bondDepository.marketPrice(_bondId);
+    // Max bond reduced by 5% to account for possible tx-failing maxBond reduction in next block
+    uint256 maxBond = maxPayout.mul(bondPrice).div(100).mul(95); // 18 decimal
+    uint256 amountToBond = (maxBond > _ids.length.mul(1e18))
+      ? _ids.length.mul(1e18)
+      : maxBond;
+    
+    // Get our vault by ID
+    address vault = nftxFactory.vault(_vaultId);
+    IERC20 vaultToken = IERC20(vault);
+
+    // Store initial balance
+    uint256 existingBalance = vaultToken.balanceOf(address(this));
 
     // Convert ERC721 to ERC20
     // The vault is an ERC20 in itself and can be used to transfer and manage
-    address vault = _mint721(vaultId, ids, to);
-    IERC20 vaultToken = IERC20(vault);
-
+    _mint721(vault, _ids, _to);
+    
     // Bond ERC20 in FloorDAO
-    vaultToken.approve(address(bondDepository), amountToBond);
-    bondDepository.deposit(bondId, amountToBond, maxPrice, to, address(0));
+    if (vaultToken.allowance(address(this), address(bondDepository)) < type(uint256).max) {
+      vaultToken.approve(address(bondDepository), type(uint256).max);
+    }
+    bondDepository.deposit(_bondId, amountToBond, _maxPrice, _to, address(0));
+
+    // Calculate remaining from initial balance and timelock
+    remaining_ = vaultToken.balanceOf(address(this)).sub(existingBalance);
+    if (remaining_ > 0) {
+      _addToTimelock(vault, remaining_, _to);
+    }
+  }
+
+  /**
+   * @notice             claim notes for user
+   * @param _user        the user to claim for
+   * @param _indexes     the note indexes to claim
+   * @return amount_     sum of amount sent in vToken
+   */
+  function claim(address _user, uint256[] memory _indexes, address _vault) external override returns (uint256 amount_) {
+    uint48 time = uint48(block.timestamp);
+
+    for (uint256 i = 0; i < _indexes.length; i++) {
+      (uint256 pay, bool matured) = pendingFor(_user, _indexes[i], _vault);
+      require(matured, "Depository: note not matured");
+      if (matured) {
+        notes[_user][_vault][_indexes[i]].claimed = time; // mark as claimed
+        amount_ += pay;
+      }
+    }
 
-    // Return remaining ERC20
-    uint256 remaining = vaultToken.balanceOf(address(this));
-    vaultToken.safeTransfer(to, remaining);
+    IERC20 vaultToken = IERC20(_vault);
+    vaultToken.safeTransfer(_user, amount_);
```

Comment:
Do we need to protect against Reentrancy here as doing a safe transfer to the User could trigger an additional claim etc?